From ed76f07c06370a229ed845000e52493f5f278139 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Thu, 23 Oct 2025 01:59:08 +0200
Subject: [PATCH] Check if running under X11 on launch

The settings are rely on X11, and cannot be used under Wayland. Instead of
crashing, just show a message dialog and exit.
---
 src/lxinput.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/lxinput.c b/src/lxinput.c
index ee62cc8..84d7cae 100644
--- a/src/lxinput.c
+++ b/src/lxinput.c
@@ -247,6 +247,24 @@ int main(int argc, char** argv)
 
     gtk_icon_theme_prepend_search_path(gtk_icon_theme_get_default(), PACKAGE_DATA_DIR);
 
+#if GTK_CHECK_VERSION(3, 0, 0)
+    /* check if not running under X11 */
+    if (!GDK_IS_X11_DISPLAY(gdk_display_get_default()))
+    {
+        GtkWidget *dialog = gtk_message_dialog_new(
+            NULL,
+            GTK_DIALOG_MODAL,
+            GTK_MESSAGE_ERROR,
+            GTK_BUTTONS_OK,
+            _("This tool can be used only under X11.")
+        );
+
+        gtk_dialog_run(GTK_DIALOG(dialog));
+        gtk_widget_destroy(dialog);
+        return 1;
+    }
+#endif
+
     /* build the UI */
     builder = gtk_builder_new();
 
